internal/currency/converter.go

```diff
@@ -0,0 +1,116 @@
+package currency
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"time"
+)
+
+const baseCurrencyAPIURL = "https://api.currencyapi.com/v3/"
+
+type ConvertResponse struct {
+	Meta Meta                 `json:"meta"`
+	Data map[string]*Currency `json:"data"`
+}
+type Meta struct {
+	LastUpdatedAt time.Time `json:"last_updated_at"`
+}
+type Currency struct {
+	Code  string  `json:"code"`
+	Value float64 `json:"value"`
+}
+
+type Converter interface {
+	Convert(amount *big.Float, currency string) (*big.Float, error)
+}
+
+type ConverterCaller interface {
+	ConvertCurrency(req map[string]string) (*ConvertResponse, error)
+}
+
+type APICurrencyConverter struct {
+	convertCaller ConverterCaller
+}
+
+func NewCurrencyConverter(caller ConverterCaller) Converter {
+	return &APICurrencyConverter{convertCaller: caller}
+}
+
+type APICurrencyConverterCaller struct {
+	client *http.Client
+	apiKey string
+}
+
+func WithAPIKey(apiKey string) APICurrencyConverterCallerOptions {
+	return func(a *APICurrencyConverterCaller) {
+		a.apiKey = apiKey
+	}
+}
+
+func WithHTTPClient(client *http.Client) APICurrencyConverterCallerOptions {
+	return func(a *APICurrencyConverterCaller) {
+		a.client = client
+	}
+}
+
+type APICurrencyConverterCallerOptions func(*APICurrencyConverterCaller)
+
+func NewAPICurrencyConverterCaller(ops ...APICurrencyConverterCallerOptions) (ConverterCaller, error) {
+	caller := &APICurrencyConverterCaller{}
+	for _, op := range ops {
+		op(caller)
+	}
+	if caller.apiKey == "" {
+		return nil, fmt.Errorf("no api key present")
+	}
+	if caller.client == nil {
+		caller.client = http.DefaultClient
+	}
+	return caller, nil
+}
+
+func (m *APICurrencyConverterCaller) ConvertCurrency(req map[string]string) (*ConvertResponse, error) {
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	httpReq, err := http.NewRequest(http.MethodGet, baseCurrencyAPIURL+"latest", bytes.NewBuffer(jsonReq))
```

Comment:
Does this actually work? You can attach a body to GET but unless I am mistaken it is ignored by the server. In my understanding the payload should be set via query params as int  `?param1=value1&....`

Reply:
You are correct that GET with body is just nonsense, but their API actually behaves like that. 

```
curl --location --request GET 'https://api.currencyapi.com/v3/latest' \                                                         INT х │ 18:21:20 
--header 'apikey: APIKEY' \
--header 'Content-Type: application/json' \
--data '{
   "base_currency": "USD",
   "currencies": "EUR"
}'
```

Returns:

```
{"meta":{"last_updated_at":"2023-09-03T23:59:59Z"},"data":{"EUR":{"code":"EUR","value":0.9275901676}}} 
```

It's just bad API design on their part, should've been a POST. Would be too complex to make it with query params + path params since it supports converting one to many currencies.